src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+                    }
+                }
+                catch (Exception ex) {
+                    Logger.Error(ex, "Error while trying to acquire a lock named {0}.", name);
+                    throw;
+                }
+
+                Logger.Debug("Could not acquire a lock named {0}.", name);
+                return false;
+            }
+        }
+
+        public IDistributedLock AcquireLock(string name, TimeSpan maxLifetime, TimeSpan timeout) {
+            IDistributedLock lockResult;
+            return TryAcquireLock(name, maxLifetime, timeout, out lockResult) ? lockResult : null;
+        }
+
+        public void DisposeLock(IDistributedLock @lock) {
+            var childLifetimeScope = CreateChildLifetimeScope(@lock.Name);
+
+            try {
+                var repository = childLifetimeScope.Resolve<IRepository<LockRecord>>();
+                var transactionManager = childLifetimeScope.Resolve<ITransactionManager>();
+                transactionManager.RequireNew(IsolationLevel.ReadCommitted);
+                var record = repository.Get(@lock.Id);
+
+                if (record != null) {
+                    if (record.ReferenceCount > 0)
+                        record.ReferenceCount--;
+                }
+            }
+            catch (Exception ex) {
+                if (ex.IsFatal()) throw;
+                Logger.Error(ex, "An non-fatal error occurred while trying to dispose a distributed lock with name {0} and ID {1}.", @lock.Name, @lock.Id);
+            }
+            finally {
+                childLifetimeScope.Dispose();
+            }
+        }
+
+        protected virtual bool TryAcquireLockRecord(string name, TimeSpan maxLifetime, out IDistributedLock @lock) {
+            @lock = null;
+            var childLifetimeScope = CreateChildLifetimeScope(name);
+
+            try {
+                var transactionManager = childLifetimeScope.Resolve<ITransactionManager>();
+                transactionManager.RequireNew(IsolationLevel.ReadCommitted);
+
+                // This way we can create a nested transaction scope instead of having the unwanted effect
+                // of manipulating the transaction of the caller.
+                var repository = childLifetimeScope.Resolve<IRepository<LockRecord>>();
+
+                // Find an existing, active lock, if any.
+                var record = repository.Table.FirstOrDefault(x => x.Name == name && (x.ValidUntilUtc >= _clock.UtcNow || x.ReferenceCount > 0));
+
+                // The current owner name (based on machine name and current thread ID).
+                var currentOwnerName = GetOwnerName();
+                var canAcquireLock = false;
+
+                // Check if there's already an active lock.
+                if (record != null) {
+                    // Check if the owner of the lock is the one trying to acquire the lock.
+                    if (record.Owner == currentOwnerName) {
+                        record.ReferenceCount++;
```

Comment:
You're right, we need NHibernate to execute the UPDATE statement as soon as possible.